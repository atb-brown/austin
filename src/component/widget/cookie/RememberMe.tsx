import React, { useCallback, useEffect } from "react";
import { useCookies } from "react-cookie";
import "../../../common.css";

const clickMeToRemember = "Click me if you want me to remember you!";
const clickMeToForget = "Welcome back! Click me if you want me to forget.";
const clickOk = "; if you're okay with that, click OK.";
const confirmRememberTxt =
  "This will require me to save a cookie on your computer" + clickOk;
const confirmForgetTxt =
  "This will remove the cookie from your computer" + clickOk;

const cookieName = "remembered";
const title =
  "This component allows the site to remember a user. It's an exercise" +
  " in handling cookies and component state!";

/**
 * This button allows the user to be remembered.
 *
 * @return {JSX.Element}
 */
export default function RememberMe(): JSX.Element {
  const [cookies, setCookie, removeCookie] = useCookies([cookieName]);
  const [remember, setRemember] = React.useState(
    Object.hasOwn(cookies, cookieName),
  );
  const [buttonTxt, setButtonTxt] = React.useState(
    remember ? clickMeToRemember : clickMeToForget,
  );
  const [confirmTxt, setConfirmTxt] = React.useState(
    remember ? confirmForgetTxt : confirmRememberTxt,
  );

  const onClick: () => void = useCallback(() => {
    if (confirm(confirmTxt)) {
      if (!remember) {
        setCookie(cookieName, "true");
      } else {
        removeCookie(cookieName);
      }
    }
  }, [confirmTxt, removeCookie, remember, setCookie]);

  useEffect(() => {
    if (Object.hasOwn(cookies, cookieName)) {
      setRemember(true);
      setButtonTxt(clickMeToForget);
      setConfirmTxt(confirmForgetTxt);
    } else {
      setRemember(false);
      setButtonTxt(clickMeToRemember);
      setConfirmTxt(confirmRememberTxt);
    }
  }, [cookies]);
  return (
    <div
      className="clickable"
      data-testid="remember-me"
      onClick={onClick}
      style={{
        alignItems: "center",
        borderRadius: 10,
        display: "flex",
        fontSize: "50%",
        height: 60,
        justifyContent: "center",
        width: "25%",
      }}
      title={title}
    >
      {buttonTxt}
    </div>
  );
}
