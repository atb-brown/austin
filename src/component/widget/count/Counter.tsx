import React from "react";

/**
 * This button is intended to count the number of times the user clicks it.
 *
 * @return {ReactElement}
 */
export default function Counter() {
  const counterButtonText = "work in progress";

  return (
    <div
      style={{
        display: "flex",
        fontSize: 20,
        backgroundColor: "steelblue",
        alignItems: "center",
        justifyContent: "center",
        width: "25%",
        height: 60,
        borderRadius: 10,
      }}
      title="This component is a work-in-progress! I'm currently iterating on it."
    >
      {counterButtonText}
    </div>
  );
}
