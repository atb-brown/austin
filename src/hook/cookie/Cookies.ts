export enum Cookies {
  Remembered = "remembered",
  Count = "count",
}

/**
 * Process something using each cookie value.
 *
 * @param {function(string):void} doThis The function that will be processed on
 *  each cookie.
 */
export function forEach(doThis: (cookie: string) => void): void {
  for (const cookie of Object.values(Cookies)) {
    doThis(cookie);
  }
}
