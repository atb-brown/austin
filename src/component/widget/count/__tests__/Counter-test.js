import Counter from "../Counter";
import "@testing-library/jest-dom";
import { screen, render, fireEvent } from "@testing-library/react";
import { Cookies as ReactCookies } from "react-cookie";
import { Cookies } from "../../../../hook/cookie/Cookies";

const cookies = new ReactCookies();

it("renders to match snapshot without cookie", () => {
  const renderResult = render(<Counter />);

  expect(renderResult.asFragment()).toMatchSnapshot();
});

it("renders to match snapshot with cookie", async () => {
  cookies.set(Cookies.Count, 12);

  const renderResult = render(<Counter />);

  const counterComp = screen.getByTestId("counter");
  expect(counterComp).toBeDefined();
  expect(counterComp).toHaveTextContent("12");

  expect(renderResult.asFragment()).toMatchSnapshot();

  cookies.remove(Cookies.Count);
});

it("click increases the count", () => {
  render(<Counter />);

  const counterComp = screen.getByTestId("counter");
  expect(counterComp).toBeDefined();

  // Before being clicked, should display 0
  expect(counterComp).toHaveTextContent("0");

  // After being clicked once, should display 1
  fireEvent.click(counterComp);
  expect(counterComp).toHaveTextContent("1");

  // Click a few more times
  fireEvent.click(counterComp);
  fireEvent.click(counterComp);
  fireEvent.click(counterComp);
  expect(counterComp).toHaveTextContent("4");
});

it("click increases the count in the cookie when user is remembered", () => {
  cookies.set(Cookies.Remembered, true);
  cookies.set(Cookies.Count, 12);
  render(<Counter />);

  const counterComp = screen.getByTestId("counter");
  expect(counterComp).toBeDefined();

  fireEvent.click(counterComp);
  fireEvent.click(counterComp);
  fireEvent.click(counterComp);
  expect(cookies.get(Cookies.Count)).toBe(15);
});
