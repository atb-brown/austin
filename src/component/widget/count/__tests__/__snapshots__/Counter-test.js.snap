// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders to match snapshot 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "backgroundColor": "steelblue",
      "borderRadius": 10,
      "display": "flex",
      "fontSize": 20,
      "height": 60,
      "justifyContent": "center",
      "width": "25%",
    }
  }
  title="This component is a work-in-progress! I'm currently iterating on it."
>
  work in progress
</div>
`;
